File: FastMines_swing/src/main/java/fmg/swing/app/serializers/AppDataSerializer.java
package fmg.swing.app.serializers;

import java.io.*;
import java.util.Objects;
import java.util.UUID;

import fmg.common.Logger;
import fmg.common.geom.Point;
import fmg.common.geom.SizeDouble;
import fmg.core.app.AProjSettings;
import fmg.core.app.ISerializer;
import fmg.core.types.model.MosaicInitData;
import fmg.swing.app.menu.EShowElement;
import fmg.swing.app.model.AppData;

/** Application data (de)serializer. For save / restore {@link AppData} */
public class AppDataSerializer implements ISerializer {

    private static final long VERSION = 1;

    private void write(AppData data, ObjectOutput out) throws IOException {
        out.writeLong(VERSION);

        MosaicInitData mid = new MosaicInitData();
        mid.setCountMines(data.getCountMines());
        mid.setMosaicType(data.getMosaicType());
        mid.setSizeField(data.getSizeField());
        new MosaicInitDataSerializer().write(mid, out);

        out.writeBoolean(data.getActiveUserId() != null);
        if (data.getActiveUserId() != null)
            out.writeUTF(data.getActiveUserId().toString());
        out.writeBoolean(data.isDoNotAskStartup());

        out.writeBoolean(data.isSystemTheme());
        for (EShowElement key : EShowElement.values())
            out.writeBoolean(data.getShowElement(key));
        out.writeBoolean(data.isZoomAlwaysMax());
        out.writeBoolean(data.isUseUnknown());
        out.writeBoolean(data.isUsePause());
        out.writeInt(data.getLocation().x);
        out.writeInt(data.getLocation().y);
        out.writeDouble(data.getSizeMosaic().width);
        out.writeDouble(data.getSizeMosaic().height);
    }

    private void read(AppData data, ObjectInput in) throws IOException {
        long ver = in.readLong();
        if (VERSION != ver)
            throw new IllegalArgumentException(AppDataSerializer.class.getSimpleName() + ": Unsupported version " + ver);

        MosaicInitData mid = new MosaicInitDataSerializer().read(in);
        data.setCountMines(mid.getCountMines());
        data.setMosaicType(mid.getMosaicType());
        data.setSizeField(mid.getSizeField());

        if (in.readBoolean())
            data.setActiveUserId(UUID.fromString(in.readUTF()));
        data.setDoNotAskStartup(in.readBoolean());

        data.setSystemTheme(in.readBoolean());
        for (EShowElement key : EShowElement.values())
            data.setShowElement(key, in.readBoolean());
        data.setZoomAlwaysMax(in.readBoolean());
        data.setUseUnknown(in.readBoolean());
        data.setUsePause(in.readBoolean());
        int x = in.readInt();
        int y = in.readInt();
        data.setLocation(new Point(x, y));
        double w = in.readDouble();
        double h = in.readDouble();
        data.setSizeMosaic(new SizeDouble(w, h));

//        Logger.info("Read: location={0}, sizeMosaic=[{1}, {2}]", location, sizeMosaicWidth, sizeMosaicHeight);
    }

    /** Load ini data from file
     * @return <b>true</b> - successful read; <b>false</b> - not exist or fail read, and set to defaults */
    public boolean load(AppData data) {
        Objects.requireNonNull(data);

        File file = getIniFile();
        if (!file.exists()) {
            data.setDefaults();
            return false;
        }

        try (InputStream is = new FileInputStream(file);
             ObjectInputStream in = new ObjectInputStream(is))
        {
            read(data, in);
            return true;
        } catch (Exception ex) {
            Logger.error("Can`t load " + AppData.class.getSimpleName(), ex);
            data.setDefaults();
            return false;
        }
    }

    public void save(AppData data) {
        try (OutputStream os = new FileOutputStream(getIniFile());
             ObjectOutputStream out = new ObjectOutputStream(os))
        {
            write(data, out);
        } catch (Exception ex) {
            Logger.error("Can`t save " + AppData.class.getSimpleName(), ex);
        }
    }

    private static File getIniFile() {
        return new File(AProjSettings.getSettingsFileName());
    }

}

File: FastMines_uwp/FastMines.Uwp.App/Serializers/AppDataSerializer.cs
using System;
using Windows.Foundation.Collections;
using Fmg.Common;
using Fmg.Core.App;
using Fmg.Core.Types.Model;
using Fmg.Uwp.App.Model;
using Fmg.Uwp.App.Presentation;

namespace Fmg.Uwp.App.Serializers {

    /// <summary> Application data (de)serializer. For save / restore <see cref="AppData"/> </summary>
    public class AppDataSerializer : ISerializer {

        /// <summary> Singleton </summary>
        public static MosaicInitData MosaicInitData { get; } = new MosaicInitData();
        /// <summary> Singleton </summary>
        public static MenuSettings MenuSettings { get; } = new MenuSettings();

        private static readonly string KEY__MENU_SETTINGS__SECTION         = nameof(MenuSettings);
        private static readonly string KEY__MENU_SETTINGS__SPLIT_PANE_OPEN = nameof(MenuSettings) + '.' + nameof(MenuSettings.SplitPaneOpen);

        public AppData Load(IPropertySet from) {
            if (from.ContainsKey(KEY__MENU_SETTINGS__SECTION)) try {
                var mid = new MosaicInitDataSerializer().Load(from);

                var compositeMosaic = (Windows.Storage.ApplicationDataCompositeValue)from[KEY__MENU_SETTINGS__SECTION];
                var isSplitPaneOpen = (bool)compositeMosaic[KEY__MENU_SETTINGS__SPLIT_PANE_OPEN];

                return new AppData() {
                    MosaicInitData = mid,
                    SplitPaneOpen = isSplitPaneOpen
                };
            } catch(Exception ex) {
                Logger.Info($"Fail load data: {ex.Message}");
                System.Diagnostics.Debug.Assert(false, ex.Message);
            }
            return new AppData();
        }

        public void Save(AppData data, IPropertySet to) {
            new MosaicInitDataSerializer().Save(data.MosaicInitData, to);
            to[KEY__MENU_SETTINGS__SECTION] = new Windows.Storage.ApplicationDataCompositeValue() {
                [KEY__MENU_SETTINGS__SPLIT_PANE_OPEN] = data.SplitPaneOpen,
            };
        }

    }

}
